submission/races.go

```diff
@@ -0,0 +1,227 @@
+type safeSubmissionSet struct {
+	results map[string]*submissionResult
+	mu      sync.Mutex
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
+func (sub *safeSubmissionSet) request(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sub.results[logURL] != nil {
+		return false
+	}
+	sub.results[logURL] = &submissionResult{}
+	return true
+}
+
+func (sub *safeSubmissionSet) has(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	return sub.results[logURL] != nil
+}
+
+func (sub *safeSubmissionSet) sumSCTsFor(URLs map[string]bool) int {
+	var matches int
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	for url := range URLs {
+		if sub.results[url] != nil && sub.results[url].sct != nil {
+			matches++
+		}
+	}
+	return matches
+}
+
+func (sub *safeSubmissionSet) setResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	sub.results[logURL] = &submissionResult{sct: sct, err: err}
+}
+
+func (sub *safeSubmissionSet) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// postInterval calculates duration for consequent call. For first parallelStart calls duration is 0, while every next one gets additional dur interval.
+func postInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+type groupState struct {
+	Name    string
+	Success bool
+}
+
+// groupRace shuffles logs within the group, submits avoiding duplicate-requests and collects responses. Upon (not)reaching required number of SCTs signals via completion channel.
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	state *safeSubmissionSet, completion chan<- groupState, submitter Submitter) {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
+	var earlyCompletion bool
+	var completionMu sync.Mutex
+	// Randomize the order in which we send requests to the logs in a group
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		go func(i int, logURL string) {
+			time.Sleep(postInterval(i, parallelStart, PostBatchInterval))
+			if ctx.Err() != nil {
+				counter <- count{}
+				return
+			}
```

Comment:
It would be better if the `Sleep` here were interruptible by a cancelled context -- something like:
```golang
	ticker := time.NewTicker(postInterval(i, parallelStart, PostBatchInterval))
	defer ticker.Stop()
	select {
	case <- ctx.Done():
		return
	case <- ticker.C:
	}
```

Reply:
I don't need a regular ticker. Added timeAfter.

Reply:
Yeah, I forgot about that -- `time.After` makes more sense.